### PR TITLE
base Visual Autocomplete

### PR DESCRIPTION
**Pull request by oshi97**
This PR tries to implement Visual Autocomplete in a more flexible
way than a huge array of config options. Externally, VisualAutocompleteSections
can be passed as children into a VisualSearchBar component. This allows the sections
to be organized in many more degrees of freedom than config options could allow, as well
as allowing users to insert whatever other random markup they desire.

Params like universalLimit and restrictedVerticals are calculated from the children passed into
the component, to try to reduce the amount of config options.

Also fixed a (likely unimportant) edge case with near me searches, and refactored some things
to make VisualSearchBar simpler.

J=SLAP-1709
TEST=manual

checked I can tab through the autocomplete options and visual entities

debouncing works as expected

loading state is applied (did not walk through it with product but should be easy to tinker with later on)

test universalLimit is passed correctly

test that searches will perform an autocomplete request to check for near me search intent when an autocomplete option
is selected. previously, only searches that were directly typed in could trigger near me behavior

### CODE REVIEW

File: sample-app/src/components/InputDropdown.tsx [hunk 1]
```diff
@@ -14,7 +14,7 @@ interface Props {
   screenReaderInstructions: string,
   screenReaderInstructionsId: string,
   screenReaderText: string,
-  onlyAllowDropdownOptionSubmissions: boolean,
+  onlyAllowDropdownOptionSubmissions?: boolean,
```
Comment (tmeyer2115) [inline sample-app/src/components/InputDropdown.tsx:17]: Out of scope of your PR, but what does this prop do? The name is somewhat confusing.

File: sample-app/src/hooks/useNearMeSearch.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+type QueryFunc = () => Promise<void>
+type AutocompleteRef = MutableRefObject<Promise<AutocompleteResponse | undefined> | undefined>
+
+/**
+ * Returns a search action that will handle near me searches, by first checking
+ * for near me intents using an autocomplete request.
+ * You can optionally use the provided ref to store autocomplete responses, to avoid
+ * making unnecessary autocomplete requests.
+ */
+export default function useNearMeSearch(
```
Comment (tmeyer2115) [inline sample-app/src/hooks/useNearMeSearch.ts:15]: This seems like something @yen-tt wanted to add in a previous PR. We ultimately decided against having a single method with all this responsibility. See the relevant comments here:
- https://github.com/yext/answers-headless-react/pull/57#discussion_r752288311
- https://github.com/yext/answers-headless-react/pull/57#discussion_r752293077

File: sample-app/src/components/SearchBar.tsx [hunk 3]
```diff
@@ -33,16 +30,12 @@ export default function SearchBar({
...
-   * Allow a query search to wait on the response to the autocomplete request right
-   * before the search execution in order to retrieve the search intents
-   */
-  const autocompletePromiseRef = useRef<Promise<AutocompleteResponse | undefined>>();
   const [autocompleteResponse, executeAutocomplete] = useSynchronizedRequest(() => {
     return isVertical
       ? answersActions.executeVerticalAutocomplete()
       : answersActions.executeUniversalAutocomplete();
   });
+  const [ executeQuery, autocompletePromiseRef ] = useNearMeSearch(answersActions, geolocationOptions);
```
Comment (tmeyer2115) [inline sample-app/src/components/SearchBar.tsx:38]: It feels strange that a method called `useNearMeSearch` is returning a ref to an auto-complete promise.

File: sample-app/src/components/VisualAutocomplete/VisualSearchBar.tsx [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+            isLoading={isLoading}
+          />
+        }
+        cssClasses={{
+          dropdownContainer: 'Autocomplete',
+          inputElement: 'SearchBar__input',
+          inputContainer: 'SearchBar__inputContainer'
+        }}
+      >
+        {options.length > 0 && renderDropdownSection()}
```
Comment (tmeyer2115) [inline sample-app/src/components/VisualAutocomplete/VisualSearchBar.tsx:118]: One thing that feels odd to me is the concept of the `VisualAutocompleteSection`. When I see something like that, I would assume we could use it to represent all pieces of the drop-down experience, including query suggestions. But, query suggestions seems to be its own special flower. Will recent searches have to be a special flower as well, or can we use a Section for that?

File: sample-app/src/components/VisualAutocomplete/VisualAutocompleteSection.tsx [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  limit?: number
+}
+
+/**
+ * VisualAutocompleteSection is responsible for providing configuration to VisualAutocompleteEntities,
+ * and allowing users to specify arbitrary VisualAutocomplete layouts.
+ * 
+ * This is done by providing results through props.children, using an FACC (function as a child component).
+ */
+export function VisualAutocompleteSection(_: VisualAutocompleteSectionProps) {
```
Comment (tmeyer2115) [inline sample-app/src/components/VisualAutocomplete/VisualAutocompleteSection.tsx:15]: I'm not totally sold on the purpose of `VisualAutocompleteSection`. It seems like not much more than a pass-through to `VisualAutocompleteEntities`. Do you anticipate it will do more in the future?

File: sample-app/src/components/VisualAutocomplete/VisualSearchBar.tsx (hunk 4 above)
Comment (tmeyer2115) [reply sample-app/src/components/VisualAutocomplete/VisualSearchBar.tsx:118]: IMO, if `VisualAutocompleteSection` can only be used for entity previews, we should just invoke the `VisualAutocompleteEntities` Component directly in the `props.children` of the `VisualSearchBar`.

File: sample-app/src/components/InputDropdown.tsx (hunk 1 above)
Comment (oshi97) [reply sample-app/src/components/InputDropdown.tsx:17]: This is from the FilterSearch pr, it's so that you can only submit options from the autocomplete dropdown, and not any arbitrary input ([I was also confused by it](https://github.com/yext/answers-headless-react/pull/64#discussion_r759509677))

**Comment by oshi97** [comment]
2 rebases later this PR has been moved this to https://github.com/yext/answers-headless-react/pull/77
